segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ChunkStorageManager.java

```diff
@@ -0,0 +1,1350 @@
+            }
+        });
+    }
+
+    /**
+     * Determines whether the given StreamSegment exists or not.
+     *
+     * @param streamSegmentName The name of the StreamSegment.
+     * @param timeout           Timeout for the operation.
+     * @return A CompletableFuture that, when completed, will contain the information requested. If the operation failed,
+     * it will contain the cause of the failure.
+     */
+    @Override
+    public CompletableFuture<Boolean> exists(String streamSegmentName, Duration timeout) {
+        checkInitialized();
+        return execute( () -> {
+            long traceId = LoggerHelpers.traceEnter(log, "exists", streamSegmentName);
+            Preconditions.checkNotNull(streamSegmentName, "streamSegmentName");
+            try (MetadataTransaction txn = metadataStore.beginTransaction()) {
+                SegmentMetadata segmentMetadata = (SegmentMetadata) txn.get(streamSegmentName);
+                val retValue = segmentMetadata == null ? false : segmentMetadata.isActive();
+                LoggerHelpers.traceLeave(log, "exists", traceId, retValue);
+                return retValue;
+            }
+        });
+    }
+
+    @Override
+    public void close() {
+        try {
+            if (null != this.metadataStore) {
+                this.metadataStore.close();
+            }
+        } catch (Exception e) {
+            log.warn("Error during close", e);
+        }
+        this.closed.set(true);
+    }
+
+    /**
+     *
+     * @param streamSegmentName
+     * @return
+     */
+    private ConcurrentSkipListMap<Long, String> getReadIndex(String streamSegmentName) {
+        ConcurrentSkipListMap<Long, String> readIndex;
+        if (cachedReadIndex.containsKey(streamSegmentName)) {
+            readIndex = cachedReadIndex.get(streamSegmentName);
+        } else {
+            val newReadIndex = new ConcurrentSkipListMap<Long, String>();
+            val oldReadIndex = cachedReadIndex.putIfAbsent(streamSegmentName, newReadIndex);
+            readIndex = null != oldReadIndex ? oldReadIndex : newReadIndex;
+        }
+        return readIndex;
+    }
+
+    /**
+     * Executes the given Callable and returns its result, while translating any Exceptions bubbling out of it into
+     * StreamSegmentExceptions.
+     *
+     * @param operation     The function to execute.
+     * @param <R>           Return type of the operation.
+     * @return CompletableFuture<R> of the return type of the operation.
+     */
+    private <R> CompletableFuture<R> execute(Callable<R> operation) {
+        return CompletableFuture.supplyAsync(() -> {
+            Exceptions.checkNotClosed(this.closed.get(), this);
+            try {
+                return operation.call();
+            } catch (Exception e) {
+                throw new CompletionException(e);
+            }
+        }, this.executor);
+    }
+
+    private String getNewChunkName(String segmentName, long offset) throws Exception {
```

Comment:
static ?

Reply:
Actually planning to include container epoch in segment chunk name. (uuids are terrible for debugging)